### PR TITLE
*: fix the bug about YEAR(0.9) returns NULL instead of 0 in NO_ZERO_DATE mode

### LINKED ISSUE #23402: min() returns wrong result
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
```mysql
mysql> CREATE TABLE `t1`  (
  `COL102` decimal(64, 4) NULL DEFAULT NULL,
  `COL103` decimal(64, 4) NULL DEFAULT NULL,
  `COL1` decimal(64, 4) GENERATED ALWAYS AS ((`COL102` / `COL103`)) STORED NULL,
  `COL2` varchar(20) CHARACTER SET utf8mb4 COLLATE utf8mb4_bin NULL DEFAULT NULL);
Query OK, 0 rows affected (0.01 sec)

mysql> insert into t1 values(21356288886702.5862, 226731074325481.5454, DEFAULT, '蝟韶蜟檊蜪櫛闢眩梘挨輟萐鰌螟橇覅f褡腲羬');
Query OK, 1 row affected (0.01 sec)

mysql> insert into t1 values(889877983770074.9164, 978025730846182.9272, DEFAULT, 'ü靔荂呍鈉滽帘厧酥厛籶呂箪嶢焳逞區賹邷菉');
Query OK, 1 row affected (0.00 sec)

mysql> insert into t1 values(-430735588411836.2316, 47557795333545.0596, DEFAULT, 'ú蓼栰絏嚹逹量眴踷娎覀詶毇魺蹏蚠侤鑏粻勁');
Query OK, 1 row affected (0.00 sec)

```

### 2. What did you expect to see? (Required)
```mysql
mysql> select year(col1), min(col2) from t1 group by year(col1);
+------------+-----------------------------------------+
| year(col1) | min(col2)                               |
+------------+-----------------------------------------+
| NULL       | ú蓼栰絏嚹逹量眴踷娎覀詶毇魺蹏蚠侤鑏粻勁 |
|          0 | ü靔荂呍鈉滽帘厧酥厛籶呂箪嶢焳逞區賹邷菉 |
+------------+-----------------------------------------+
2 rows in set (0.07 sec)

```
### 3. What did you see instead (Required)
```mysql
mysql> select year(col1), min(col2) from t1 group by year(col1);
+------------+-----------------------------------------+
| year(col1) | min(col2)                               |
+------------+-----------------------------------------+
|          0 | 蝟韶蜟檊蜪櫛闢眩梘挨輟萐鰌螟橇覅f褡腲羬 |
| NULL       | ú蓼栰絏嚹逹量眴踷娎覀詶毇魺蹏蚠侤鑏粻勁 |
+------------+-----------------------------------------+
2 rows in set (0.08 sec)

```
### 4. What is your TiDB version? (Required)
```mysql
Release Version: v4.0.0-beta.2-2393-g3cfb984
Edition: Community
Git Commit Hash: 3cfb984211901c3416ec539658ac86d2cb64e81f
Git Branch: master
UTC Build Time: 2021-03-18 02:41:26
GoVersion: go1.13.5
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
```

### PR DESCRIPTION
**Pull request by jyz0309**
Signed-off-by: jyz0309 <45495947@qq.com>

### What problem does this PR solve?

Issue Number: close #23402  

Problem Summary:
YEAR(0.9) returns NULL instead of 0 in NO_ZERO_DATE mode
### What is changed and how it works?

What's Changed:
Change the behaviour when NO_ZERO_DATE is turn on, the func `handleInvalidZeroTime` will return false.

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Manual test (add detailed scripts or steps below)

### Release note 

- fix the bug about  NO_ZERO_DATE mode

### CODE REVIEW

File: expression/builtin_time.go [hunk 1]
```diff
@@ -7111,5 +7111,5 @@ func handleInvalidZeroTime(ctx sessionctx.Context, t types.Time) (bool, error) {
 	if isOriginalIntOrDecimalZero && !isOriginalStringZero {
 		return false, nil
 	}
-	return true, handleInvalidTimeError(ctx, types.ErrWrongValue.GenWithStackByArgs(types.DateTimeStr, t.String()))
+	return false, handleInvalidTimeError(ctx, types.ErrWrongValue.GenWithStackByArgs(types.DateTimeStr, t.String()))
```
Comment (djshow832) [inline expression/builtin_time.go:7114]: I saw some built-in functions (such as DayOfYear) also called `handleInvalidZeroTime` without checking `NO_ZERO_DATE`. Is the result still expected in this case?

Comment (jyz0309) [reply expression/builtin_time.go:7114]: I know. And actually I am confused about here. I don't know why we need to return NULL when year(0.9) but return 0 when year(0)

**Pull request closed without merging**